src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractToMethodConverter.cs

```diff
@@ -0,0 +1,203 @@
+                    editor.RemoveNode(nodeToRemoveIfFollowedByReturn);
+                }
+                else
+                {
+                    leadingTrivia = Helpers.GetTrivia(replacingExpression);
+                }
+
+                editor.ReplaceNode(replacingExpression, CreateInvocationExpression(queryExpression).WithComments(leadingTrivia, _trivia).KeepCommentsAndAddElasticMarkers());
+                editor.RemoveNode(ForEachInfo.ForEachStatement);
+            }
+
+            SyntaxTrivia[] GetTriviaFromVariableDeclarator(VariableDeclaratorSyntax variableDeclarator)
+                => Helpers.GetTrivia(variableDeclarator.Identifier, variableDeclarator.Initializer.EqualsToken, variableDeclarator.Initializer.Value);
+
+            SyntaxTrivia[] GetTriviaFromNode(SyntaxNode node)
+            {
+                switch (node.Kind())
+                {
+                    case SyntaxKind.LocalDeclarationStatement:
+
+                        var localDeclaration = (LocalDeclarationStatementSyntax)node;
+                        if (localDeclaration.Declaration.Variables.Count != 1)
+                        {
+                            throw ExceptionUtilities.Unreachable;
+                        }
+
+                        return new IEnumerable<SyntaxTrivia>[] {
+                            Helpers.GetTrivia(localDeclaration.Declaration.Type),
+                            GetTriviaFromVariableDeclarator(localDeclaration.Declaration.Variables[0]),
+                            Helpers.GetTrivia(localDeclaration.SemicolonToken)}.SelectMany(x => x).ToArray();
+
+                    case SyntaxKind.VariableDeclarator:
+                        return GetTriviaFromVariableDeclarator((VariableDeclaratorSyntax)node);
+
+                    case SyntaxKind.ExpressionStatement:
+                        if (((ExpressionStatementSyntax)node).Expression is AssignmentExpressionSyntax assignmentExpression)
+                        {
+                            return Helpers.GetTrivia(assignmentExpression.Left, assignmentExpression.OperatorToken, assignmentExpression.Right);
+                        }
+
+                        break;
+                }
+
+                throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        // query => query.Method()
+        // like query.Count() or query.ToList()
+        protected InvocationExpressionSyntax CreateInvocationExpression(QueryExpressionSyntax queryExpression)
+            => SyntaxFactory.InvocationExpression(
+                    SyntaxFactory.MemberAccessExpression(
+                        SyntaxKind.SimpleMemberAccessExpression,
+                        SyntaxFactory.ParenthesizedExpression(queryExpression),
+                        SyntaxFactory.IdentifierName(MethodName))).WithAdditionalAnnotations(Formatter.Annotation);
+
+        private static StatementSyntax FindPreviousStatementInBlock(StatementSyntax statement)
```

Comment:
these seems generic enough to be in StatementSyntaxExtensions.cs ?

Reply:
indeed.  and i think we have this code elsewhere... so it might be nice to have an extension.

Reply:
Thank you! We even already have GetPreviousStatement and GetNextStatement, and they work better: they support statements within switch cases not only in blocks.

---
In reply to: [195842031](https://github.com/dotnet/roslyn/pull/27130#discussion_r195842031) [](ancestors = 195842031)